authutils/validate.go

```diff
@@ -0,0 +1,219 @@
+package authutils
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+// `fence`-specific: the `pur` field indicates the purpose for the token, which
+// may also be validated. If used, it must take one of these values.
+var ALLOWED_PURPOSES []string = []string{"id", "access", "refresh", "session", "api_key"}
+
+// JWTApplication stores the state for an application needing to validate JWTs.
+type JWTApplication struct {
+	// KeyKeys for looking up RSA public keys by ID for token validation.
+	Keys *KeysManager
+}
+
+// NewJWTApplication initializes a new application.
+func NewJWTApplication(jwkURL string) *JWTApplication {
+	keysManager := NewKeysManager(jwkURL)
+	return &JWTApplication{
+		Keys: &keysManager,
+	}
+}
+
+// Decode takes an encoded token, finds the key suitable for validating this
+// token, and returns the decoded token claims.
+//
+// NOTE that this does NOT validate the claims, only the signature.
+func (application *JWTApplication) Decode(encodedToken EncodedToken) (*Claims, error) {
+	decodedToken, err := jwt.ParseSigned(encodedToken)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the key used to sign this token.
+	//
+	// There will be multiple headers in the case of multiple signatures. This
+	// doesn't apply to our use case, so just take the key ID from the first
+	// (and what should be the only) header.
```

Comment:
Its probably worth adding an if/assert for this case if we don't believe it happens so that we can track if it ever does in the wild?

Reply:
Will add

Reply:
Fixed in 0c7197e